### PR TITLE
[utils] Handle user:pass in URLs

### LINKED ISSUE #18276: Vimeo embed-only video behind basic auth
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like this: `[x]`)
- Use the *Preview* tab to see what your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2018.11.18*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2018.11.18**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### Description of your *issue*, suggested solution and other information

I'm trying to download an embed-only video, hosted on Vimeo and embedded on a page protected by [basic auth](https://en.wikipedia.org/wiki/Basic_access_authentication). I have the username and password and am able to access the page and play the video normally on a browser. I've tried the following approaches, all of which have resulted in errors.

#### 1. Using the video embed URL

```
$ youtube-dl -v https://player.vimeo.com/video/VIDEOID
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'https://player.vimeo.com/video/VIDEOID']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.11.18
[debug] Python version 2.7.15 (CPython) - Darwin-17.7.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 4.1, ffprobe 4.1
[debug] Proxy map: {}
[vimeo] VIDEOID: Downloading webpage
ERROR: Cannot download embed-only video without embedding URL. Please call youtube-dl with the URL of the page that embeds this video.
Traceback (most recent call last):
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/vimeo.py", line 488, in _real_extract
    webpage, urlh = self._download_webpage_handle(request, video_id)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 638, in _download_webpage_handle
    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query, expected_status=expected_status)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 623, in _request_webpage
    raise ExtractorError(errmsg, sys.exc_info()[2], cause=err)
ExtractorError: Unable to download webpage: HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 792, in extract_info
    ie_result = ie.extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 508, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/vimeo.py", line 503, in _real_extract
    expected=True)
ExtractorError: Cannot download embed-only video without embedding URL. Please call youtube-dl with the URL of the page that embeds this video.
```

#### 2. Using the URL of the page that embeds the video

```
$ youtube-dl -v http://PROTECTED-PAGE.EXAMPLE.com
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'http://PROTECTED-PAGE.EXAMPLE.com']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.11.18
[debug] Python version 2.7.15 (CPython) - Darwin-17.7.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 4.1, ffprobe 4.1
[debug] Proxy map: {}
[generic] PROTECTED-PAGE.EXAMPLE: Requesting header
WARNING: Could not send HEAD request to http://PROTECTED-PAGE.EXAMPLE.com: HTTP Error 401: Authorization Required
[generic] PROTECTED-PAGE.EXAMPLE: Downloading webpage
ERROR: Unable to download webpage: HTTP Error 401: Authorization Required (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 605, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2211, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

#### 3. Adding `--username` and `--password` options to the command line

```
$ youtube-dl -v --username THEUSERNAME --password THEPASSWORD http://PROTECTED-PAGE.EXAMPLE.com
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'--username', u'PRIVATE', u'--password', u'PRIVATE', u'http://PROTECTED-PAGE.EXAMPLE.com']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.11.18
[debug] Python version 2.7.15 (CPython) - Darwin-17.7.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 4.1, ffprobe 4.1
[debug] Proxy map: {}
[generic] PROTECTED-PAGE.EXAMPLE: Requesting header
WARNING: Could not send HEAD request to http://PROTECTED-PAGE.EXAMPLE.com: HTTP Error 401: Authorization Required
[generic] PROTECTED-PAGE.EXAMPLE: Downloading webpage
ERROR: Unable to download webpage: HTTP Error 401: Authorization Required (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 605, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2211, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

#### 4. Adding `username:password` to the page's URL

```
$ youtube-dl -v http://THEUSERNAME:THEPASSWORD@PROTECTED-PAGE.EXAMPLE.com
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'http://THEUSERNAME:THEPASSWORD@PROTECTED-PAGE.EXAMPLE.com']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.11.18
[debug] Python version 2.7.15 (CPython) - Darwin-17.7.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 4.1, ffprobe 4.1
[debug] Proxy map: {}
[generic] THEUSERNAME:THEPASSWORD@PROTECTED-PAGE.EXAMPLE: Requesting header
WARNING: Could not send HEAD request to http://THEUSERNAME:THEPASSWORD@PROTECTED-PAGE.EXAMPLE.com: nonnumeric port: 'THEPASSWORD@PROTECTED-PAGE.EXAMPLE.com'
[generic] THEUSERNAME:THEPASSWORD@PROTECTED-PAGE.EXAMPLE: Downloading webpage
ERROR: Unable to download webpage: nonnumeric port: 'THEPASSWORD@PROTECTED-PAGE.EXAMPLE.com' (caused by InvalidURL(u"nonnumeric port: 'THEPASSWORD@PROTECTED-PAGE.EXAMPLE.com'",)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 605, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2211, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/local/bin/youtube-dl/youtube_dl/utils.py", line 983, in http_open
    req)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 1167, in do_open
    h = http_class(host, timeout=req.timeout, **http_conn_args)
  File "/usr/local/bin/youtube-dl/youtube_dl/utils.py", line 882, in _create_http_connection
    hc = http_class(*args, **compat_kwargs(kwargs))
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 736, in __init__
    (self.host, self.port) = self._get_hostport(host, port)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 777, in _get_hostport
    raise InvalidURL("nonnumeric port: '%s'" % host[i+1:])
```

### LINKED ISSUE #20258: Support for Downloading HTTP Authentication Protected Podcast Feeds
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like this: `[x]`)
- Use the *Preview* tab to see what your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2019.03.01*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2019.03.01**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [ ] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [x] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### Description of your *issue*, suggested solution and other information

One of my favorite uses of youtube-dl is as an archiving solution. Recently I was attempting to archive some podcasts that were protected by HTTP Authentication (https://username:password@example.com/podcast/feed). I know that premium podcasts distribute their URLs in this format for supporters. Most podcast clients can parse this type of url.

```Powershell
PS C:\Archive\Podcasts> .\youtube-dl.exe https://username:password@example.com/podcast/feed
[generic] feed: Requesting header
WARNING: Could not send HEAD request to https://username:password@example.com/podcast/feed: nonnumeric port: 'password@example.com'
[generic] feed: Downloading webpage
ERROR: Unable to download webpage: nonnumeric port: 'password@example.com' (caused by InvalidURL("nonnumeric port: 'password@example.com'",)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```
I tried every combination of the `-u username -p password` I could.

#### Solution

You can get around this by inserting the Authorization header into youtube-dl's request.

The authorization header is formatted as "Authorization: Basic [username:password encoded with base64]"

Note that you have to encode your username:password beforehand.

username:password encoded with Base64 is dXNlcm5hbWU6cGFzc3dvcmQ=

So your authorization header will look like this `Authorization: Basic dXNlcm5hbWU6cGFzc3dvcmQ=`

You can insert this header with the `--add-header` command

Your new youtube-dl command will look like this:

```powershell
PS C:\Archive\Podcasts> .\youtube-dl.exe https://example.com/podcast/feed --add-header "Authorization: Basic dXNlcm5hbWU6cGFzc3dvcmQ="
```
I was able to confirm that this works for me. Could this process be automated when "username:password@example.com" is detected in the URL?

### LINKED ISSUE #26211: error when entering username and password
## Checklist

- [x] I'm reporting a broken site support issue
- [x ] I've verified that I'm running youtube-dl version **2020.07.28**
- [x ] I've checked that all provided URLs are alive and playable in a browser
- [ x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x ] I've searched the bugtracker for similar bug reports including closed ones
- [x ] I've read bugs section in FAQ

## Verbose log

```
PASTE VERBOSE LOG HERE
```
[tim@cabra ~]$ youtube-dl --version
2020.07.28
[tim@cabra ~]$ youtube-dl -o '%(title)s.%(ext)s'   'http://mooo.peelwiki.com/dl/BillfromNorthWales/01%20John%20Peel/1978-12-26%20John%20Peel%20BBC%20Radio%201.mp3' --username 'peel' --password 'group' --verbose
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-o', u'%(title)s.%(ext)s', u'http://mooo.peelwiki.com/dl/BillfromNorthWales/01%20John%20Peel/1978-12-26%20John%20Peel%20BBC%20Radio%201.mp3', u'--username', u'PRIVATE', u'--password', u'PRIVATE', u'--verbose']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2020.07.28
[debug] Python version 2.7.17 (CPython) - Linux-5.3.11-100.fc29.x86_64-x86_64-with-fedora-29-Twenty_Nine
[debug] exe versions: ffmpeg 4.0.5, ffprobe 4.0.5
[debug] Proxy map: {}
[generic] 1978-12-26 John Peel BBC Radio 1: Requesting header
WARNING: Could not send HEAD request to http://mooo.peelwiki.com/dl/BillfromNorthWales/01%20John%20Peel/1978-12-26%20John%20Peel%20BBC%20Radio%201.mp3: HTTP Error 401: Authorization Required
[generic] 1978-12-26 John Peel BBC Radio 1: Downloading webpage
ERROR: Unable to download webpage: HTTP Error 401: Authorization Required (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 627, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2238, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib64/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib64/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib64/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)

[tim@cabra ~]$ 

## Description

WRITE DESCRIPTION HERE

This is the first time I have had to enter a username/password combo into youtube-dl. From the browser I enter this url followed by username/password and it opens up a media player within the browser. This is the next level url

https://peel.fandom.com/wiki/1978

search for 26 December 1978 and then head to the bottom of that page for the mooo links.

I have tried a few different options with the same results - http and https. Username/password in single/double and no quotes with the same response

Regards

### PR DESCRIPTION
**Pull request by hhirtz**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Fixes "nonnumeric port" errors when youtube-dl is given URLs with
usernames and passwords such as:

    http://username:password@example.com/myvideo.mp4

Refs:
- https://en.wikipedia.org/wiki/Basic_access_authentication
- https://tools.ietf.org/html/rfc1738#section-3.1
- https://docs.python.org/3.8/library/urllib.parse.html#urllib.parse.urlsplit

Fixes #18276 (point 4)
Fixes #20258
Fixes #26211 (see comment)

### CODE REVIEW

**Comment by jeanthom** [comment]
LGTM

**Comment by hhirtz** [comment]
Simplified the code and ticked the "improvement" box instead of the "bug fix" one

**Comment by hhirtz** [comment]
rebased the branch

**Comment by jeanthom** [comment]
still LGTM

**Comment by jeanthom** [comment]
Dear maintainers, please take some time to consider this issue, it finally adds a very important functionality to youtube-dl!

**Review by dirkf — CHANGES_REQUESTED** [review]
Let's finally add this, but with a structure matching yt-dlp.

File: youtube_dl/utils.py [hunk 1]
```diff
@@ -2182,6 +2182,21 @@ def sanitize_url(url):
...
+        netloc = parts.hostname
+        if parts.port is not None:
+            netloc = parts.hostname + ':' + parts.port
+        parts = parts._replace(netloc=netloc)
+        url = compat_urlparse.urlunsplit(parts)
+    return url, username, password
+
+
 def sanitized_Request(url, *args, **kwargs):
     return compat_urllib_request.Request(escape_url(sanitize_url(url)), *args, **kwargs)
```
Comment (dirkf) [inline youtube_dl/utils.py:2201]:
```suggestion
    url, auth_header = extract_basic_auth(escape_url(sanitize_url(url)))
    if auth_header is not None:
        headers = args[1] if len(args) > 1 else kwargs.get('headers')
        headers = headers or {}
        headers['Authorization'] = auth_header
        if len(args) <= 1 and kwargs.get('headers') is None:
            kwargs['headers'] = headers
            kwargs = compat_kwargs(kwargs)
    return compat_urllib_request.Request(url, *args, **kwargs)
```

**Review by dirkf — APPROVED** [review]

**Pull request merged**